src/integer/mat_z/arithmetic/div.rs

```diff
@@ -0,0 +1,292 @@
+    /// If this is not the case, the result can contain arbitrary values
+    /// which can depend on the location in memory.
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::integer::{MatZ, Z};
+    /// use std::str::FromStr;
+    ///
+    /// let mut mat1 = MatZ::from_str("[[3,6],[9,27]]").unwrap();
+    /// let integer = Z::from(3);
+    ///
+    /// let mat2 = unsafe{ mat1.div_exact(&integer)};
+    /// ```
+    pub unsafe fn div_exact(mut self, divisor: impl Into<Z>) -> MatZ {
+        let divisor: Z = divisor.into();
+
+        fmpz_mat_scalar_divexact_fmpz(&mut self.matrix, &self.matrix, &divisor.value);
+
+        self
+    }
+
+    /// Implements division for a [`MatZ`] matrix by a [`Z`] integer.
+    ///
+    /// Parameters:
+    /// - `divisor`: specifies the divisor by which the matrix is divided
+    ///
+    /// Returns the product of `self` and `other` as a [`MatZ`] or panics if
+    /// the divisor is 0.
+    ///
+    /// # Safety
+    /// The divisor MUST exactly divide each element in the matrix.
+    /// If this is not the case, the result can contain arbitrary values
+    /// which can depend on the location in memory.
+    ///
+    /// # Examples
+    /// ```
+    /// use qfall_math::integer::{MatZ, Z};
+    /// use std::str::FromStr;
+    ///
+    /// let mat1 = MatZ::from_str("[[3,6],[9,27]]").unwrap();
+    /// let integer = Z::from(3);
+    ///
+    /// let mat2 = unsafe{&mat1.div_exact_ref(&integer)};
+    /// ```
+    #[allow(clippy::missing_safety_doc)]
+    pub unsafe fn div_exact_ref(&self, divisor: impl Into<Z>) -> MatZ {
+        let divisor: Z = divisor.into();
+
+        let mut out = MatZ::new(self.get_num_rows(), self.get_num_columns());
+        fmpz_mat_scalar_divexact_fmpz(&mut out.matrix, &self.matrix, &divisor.value);
+
+        out
+    }
+}
+
+#[cfg(test)]
+mod test_div_exact {
+    use super::*;
+    use std::str::FromStr;
+
+    /// Checks if matrix division works fine for both borrowed.
+    #[test]
+    fn borrowed_correctness() {
+        let mat1 = MatZ::from_str("[[6,3],[3,6]]").unwrap();
+        let mat2 = mat1.clone();
+        let integer = Z::from(3);
+
+        let mat1 = unsafe { mat1.div_exact_ref(&integer) };
+        let mat2 = unsafe { mat2.div_exact(&integer) };
+
+        let mat3 = MatZ::from_str("[[2,1],[1,2]]").unwrap();
+
+        assert_eq!(mat3, mat1);
+        assert_eq!(mat3, mat2);
+    }
+
+    /// Checks if matrix division works fine for both owned.
+    #[test]
+    fn owned_correctness() {
+        let mat1 = MatZ::from_str("[[6,3],[3,6]]").unwrap();
+        let mat2 = mat1.clone();
+        let integer1 = Z::from(3);
+        let integer2 = Z::from(3);
+
+        let mat1 = unsafe { mat1.div_exact_ref(integer1) };
+        let mat2 = unsafe { mat2.div_exact(integer2) };
+
+        let mat3 = MatZ::from_str("[[2,1],[1,2]]").unwrap();
+
+        assert_eq!(mat3, mat1);
+        assert_eq!(mat3, mat2);
```

Comment:
Move assert logic together, rename `mat3` to a more meaningful `mat_cmp`.
Also applies to following tests.